komodo/release_cleanup.py

```diff
@@ -0,0 +1,195 @@
+import os
+import sys
+from argparse import ArgumentError, ArgumentParser, ArgumentTypeError
+from komodo import load_yaml, write_to_file, prettified_yaml
+
+
+def load_all_releases(files):
+
+    used_versions = {}
+
+    for filename in files:
+        current_release = load_yaml(filename)
+
+        for lib, version in current_release.items():
+            if lib in used_versions:
+                used_versions[lib].append(version)
+            else:
+                used_versions[lib] = [version]
+
+    return used_versions
+
+
+def find_unused_versions(used_versions, repository):
+    unused_versions = {}
+    for lib, versions in repository.items():
+        for version in versions:
+            if lib in used_versions and version in used_versions[lib]:
+                continue
+
+            if lib in unused_versions:
+                unused_versions[lib].append(version)
+            else:
+                unused_versions[lib] = [version]
+
+    return unused_versions
+
+
+def remove_unused_versions(repository, unused_versions):
+    for lib, versions in unused_versions.items():
+        for version in versions:
+            repository[lib].pop(version)
+        if len(repository[lib]) == 0:
+            repository.pop(lib)
+
+
+def _is_yml(path):
+    _, ext = os.path.splitext(path)
+    return ext == ".yml"
+
+
+def _get_yml_files(path_name):
+    current_path = os.path.realpath(path_name)
+    files = os.listdir(path_name)
+    yml_files = []
+    for filename in files:
+        file_path = os.path.join(current_path, filename)
+
+        if not _is_yml(file_path):
+            continue
+        yml_files.append(file_path)
+    return yml_files
+
+
+def _valid_path_or_files(path):
+
+    yml_files = []
+
+    full_path = os.path.realpath(path)
+    if os.path.isdir(full_path):
+        yml_files += _get_yml_files(full_path)
+    elif os.path.isfile(full_path) and _is_yml(full_path):
+        yml_files.append(full_path)
+    else:
+        raise ArgumentTypeError("{} is not a valid yml-file or folder".format(path))
+    return yml_files
+
+
+def add_cleanup_parser(subparsers):
+    cleanup_parser = subparsers.add_parser(
+        "cleanup",
+        description="Clean up unused versions the repository file based on a set of releases",
+    )
+    cleanup_parser.set_defaults(func=run_cleanup)
+    cleanup_parser.add_argument(
+        "--check",
+        action="store_true",
+        help="Only print out unused library versions",
+        required=False,
+    )
+    cleanup_parser.add_argument(
+        "--repository",
+        type=load_yaml,
+        help="a yml file defining all libraries and their versions",
+    )
+    cleanup_parser.add_argument(
+        "--releases",
+        type=_valid_path_or_files,
+        help="list of release files or folders containing releases",
+        nargs="+",
+    )
+    cleanup_parser.add_argument(
```

Comment:
Please output `--output` to `STDOUT` if not specified.

And output from `--check` should be suppressed if not explicitly asked for? If I provide `--output` I still get `--check` output on `STDOUT`.

Reply:
stdout option added